tools/reduce.py

```diff
@@ -0,0 +1,303 @@
+    "-iwithprefixbefore", "--include-with-prefix-before",
+    "-iwithsysroot",
+    "--libomptarget-amdgpu-bc-path", "--libomptarget-amdgcn-bc-path",
+    "--libomptarget-nvptx-bc-path",
+    "--nvptx-arch-tool",
+    "--ptxas-path",
+    "--rocm-path",
+    "-stdlib\+\+-isystem",
+    "--system-header-prefix", "--no-system-header-prefix"
+]
+# fmt: on
+
+
+class CompdbEntry:
+    filepath: pathlib.Path
+    directory: pathlib.Path
+    arguments: List[str]
+
+    def __init__(self, entry):
+        self.filepath = pathlib.Path(entry["file"])
+        self.directory = pathlib.Path(entry["directory"])
+        try:
+            self.arguments = shlex.split(entry["command"])
+        except KeyError:
+            self.arguments = entry(["arguments"])
+
+    def change_tu_filepath(self, new_path: str):
+        old_path = str(self.filepath)
+        for i, arg in enumerate(self.arguments):
+            if old_path in arg:
+                self.arguments[i] = arg.replace(old_path, new_path)
+        self.filepath = pathlib.Path(new_path)
+
+    def to_dict(self):
+        return {
+            "directory": str(self.directory),
+            "file": str(self.filepath),
+            "arguments": self.arguments,
+        }
+
+
+def run_preprocessor_only(entry: CompdbEntry, preprocessed_tu_path: pathlib.Path):
+    args = copy.deepcopy(entry.arguments)
+    args += ["-E", "-o", str(preprocessed_tu_path)]
+    subprocess.run(args, cwd=entry.directory).check_returncode()
+
+
+def check_scip_clang_output(
+    scip_clang: pathlib.Path, compdb_path: str, pattern: re.Pattern
+) -> bool:
+    """Check if the stdout or stderr has a match for the pattern"""
+    result = subprocess.run(
+        [str(scip_clang), "--compdb-path", compdb_path, "--worker-mode=compdb"],
+        capture_output=True,
+        encoding="utf8",
+    )
+    match = re.search(pattern, result.stdout)
+    if match is not None:
+        return True
+    match = re.search(pattern, result.stderr)
+    return match is not None
+
+
+@dataclass
+class CReduce:
+    entry: CompdbEntry
+    scip_clang: pathlib.Path
+    project_root: pathlib.Path
+    scip_clang_output_pattern: re.Pattern
+    extra_args: List[str]
+    # path_flag_pattern: re.Pattern
+
+    def __post_init__(self):
+        # Lenient behavior instead of hard-coding which flags allow =
+        flags = [flag + "=?" for flag in PATH_FLAGS]
+        self.path_flag_pattern = re.compile(
+            "(?P<flag>{})(?P<path>.*)".format("|".join(flags))
+        )
+
+    def run(self, start_tu_path: pathlib.Path):
+        assert start_tu_path.is_absolute()
+        absolute_start_tu_path = start_tu_path
+        start_tu_path = start_tu_path.relative_to(self.project_root)
+        assert not start_tu_path.is_absolute()
+
+        original_tu_path = str(self.entry.filepath)
+        for i, arg in enumerate(self.entry.arguments):
+            if original_tu_path in arg:
+                self.entry.arguments[i] = str(start_tu_path)
```

Comment:
Similar to the feedback above, this could be more direct by simply doing the operation on all the arguments. Replace is a no-op if the search string is not found.

Reply:
I don't think that applies here, because we're using assignment rather than `replace`.

Specifically, the current code will replace both an absolute path and a relative path with a relative path. If I used `replace` here, then an absolute path will get mapped to an absolute path, but we do not want to use an absolute path for the main TU in the argument list, since the temporary directories created by C-Reduce may change.